api/desecapi/models.py

```diff
@@ -216,80 +216,87 @@ def write_rrsets(self, datas):
-        RRset.objects.filter(q_update | q_delete, domain=self).select_for_update()
-
-        # Figure out which RRsets are unchanged and can be excluded
-        exclude_from_update = []
-        qs_update = RRset.objects.filter(q_update, domain=self)
-        for rrset_old in qs_update.prefetch_related('records').all():
-            rrset_new = rrsets_index[(rrset_old.subname, rrset_old.type)]
-            if rrset_old.ttl != rrset_new.ttl:
-                continue
-            rrs_new = {rr.content for rr in rrsets[rrset_new]}
-            rrs_old = {rr.content for rr in rrset_old.records.all()}
-            if rrs_new != rrs_old:
-                continue
-            # Old and new contents do not differ, so we can skip this RRset
-            del rrsets[rrset_new]
-            exclude_from_update.append(rrset_old)
-
-        # Do not process new RRsets that are empty (and did not exist before)
-        # This is to avoid unnecessary pdns requests like (A: ...; AAAA: None)
-        qs_delete = RRset.objects.filter(q_delete, domain=self)
-        qs_delete_values = qs_delete.values_list('subname', 'type')
-        # We modify the rrsets dictionary and thus loop over a copy of it
-        for rrset, rrs in list(rrsets.items()):
-            if rrs or (rrset.subname, rrset.type) in qs_delete_values:
-                continue
-            # RRset up for deletion does not exist
-            del rrsets[rrset]
-
-        # Clear or delete RRsets
-        RR.objects.filter(rrset__in=qs_update).exclude(rrset__in=exclude_from_update).delete()
-        RRset.objects.filter(q_delete, domain=self).delete()
-
-        # Prepare and save new RRset contents
-        # We modify the rrsets dictionary and thus loop over a copy of it
-        for rrset, rrs in list(rrsets.items()):
-            if not rrs:
-                continue
-            # (Create and) get correct RRset and update dictionary accordingly
-            del rrsets[rrset]
-            (rrset, _) = RRset.objects.get_or_create(domain=self,
-                                                     subname=rrset.subname,
-                                                     type=rrset.type,
-                                                     ttl=rrset.ttl)
-            rrsets[rrset] = [RR(rrset=rrset, content=rr.content) for rr in rrs]
-
-        RR.objects.bulk_create([rr for rrs in rrsets.values() for rr in rrs])
-
-        # Send changed RRsets to pdns
-        if rrsets and not self.owner.locked:
-            pdns.set_rrsets(self, rrsets)
+                rrsets_to_write[rrset] = []
+                q_empty |= q
+
+        # Construct querysets representing RRsets that do (not) have RR
+        # contents and lock them
+        qs_meaty = rrset_qs.filter(q_meaty).select_for_update()
+        qs_empty = rrset_qs.filter(q_empty).select_for_update()
+
+        # For existing RRsets, execute TTL updates and mark for RR update
+        rrsets_same_rrs = []
+        for rrset in qs_meaty.all():
```

Comment:
Can this loop also include the "new, non-empty" case from below (line 280ff)? Then the "todo-list pattern" makes full sense. ("for item in todo")

Reply:
No, `qs_meaty` is a queryset, i.e. it contains stuff that is currently from the database, and thus it can't contain any new RRsets.